File: Plugins/AudioPlayer/src/android/app/src/main/java/com/shortsands/audioplayer/AudioTOCTestament.java
package com.shortsands.audioplayer;

/**
 * Created by garygriswold on 8/30/17.
 */
import android.util.Log;
import java.util.HashMap;
import java.util.Iterator;

class AudioTOCTestament {

    private static String TAG = "AudioTOCTestament";

    AudioTOCBible bible;
    String damId;
    String dbpLanguageCode;
    String dbpVersionCode;
    String collectionCode;
    String mediaType;
    HashMap<String, AudioTOCBook> booksById;
    HashMap<Integer, AudioTOCBook> booksBySeq;

    AudioTOCTestament(AudioTOCBible bible, AudioSqlite3 database, String[] dbRow) {
        this.bible = bible;
        this.booksById = new HashMap<String, AudioTOCBook>();
        this.booksBySeq = new HashMap<Integer, AudioTOCBook>();
        this.damId = dbRow[0];
        this.collectionCode = dbRow[1];
        this.mediaType = dbRow[2];
        this.dbpLanguageCode = dbRow[3];
        this.dbpVersionCode = dbRow[4];

        String query = "SELECT bookId, bookOrder, numberOfChapters" +
            " FROM AudioBook" +
            " WHERE damId = ?" +
            " ORDER BY bookOrder";
        String[] values = new String[1];
        values[0] = this.damId;
        try {
            String[][] resultSet = database.queryV1(query, values);
            for (int i=0; i<resultSet.length; i++) {
                String[] row = resultSet[i];
                AudioTOCBook book = new AudioTOCBook(this, row);
                this.booksById.put(book.bookId, book);
                this.booksBySeq.put(book.sequence, book);
            }
        } catch (Exception err) {
            Log.d(TAG,"ERROR " + AudioSqlite3.errorDescription(err));
        }
    }

    //deinit {
    //    print("***** Deinit TOCAudioTOCBible *****")
    //}

    AudioReference nextChapter(AudioReference ref) {
        AudioTOCBook book = ref.tocAudioBook;
        if (ref.chapterNum() < book.numberOfChapters) {
            int next = ref.chapterNum() + 1;
            return AudioReference.factory(ref.tocAudioBook, next, ref.fileType);
        } else {
            AudioTOCBook nextBook = this.booksBySeq.get(ref.sequenceNum() + 1);
            if (nextBook != null) {
                return AudioReference.factory(nextBook, 1, ref.fileType);
            }
        }
        return null;
    }

    AudioReference priorChapter(AudioReference ref) {
        int prior = ref.chapterNum() - 1;
        if (prior > 0) {
            return AudioReference.factory(ref.tocAudioBook, prior, ref.fileType);
        } else {
            AudioTOCBook priorBook = this.booksBySeq.get(ref.sequenceNum() - 1);
            if (priorBook != null) {
                return AudioReference.factory(priorBook, priorBook.numberOfChapters, ref.fileType);
            }
        }
        return null;
    }

    String getBookList() {
        StringBuffer array = new StringBuffer();
        Iterator<AudioTOCBook> it = this.booksBySeq.values().iterator();
        while(it.hasNext()) {
            array.append(",");
            array.append(it.next().bookId);
        }
        return array.toString();
    }

    public String toString() {
        String str = "damId=" + this.damId +
                "\n languageCode=" + this.dbpLanguageCode +
                "\n versionCode=" + this.dbpVersionCode +
                "\n mediaType=" + this.mediaType +
                "\n collectionCode=" + this.collectionCode;
        return str;
    }
}

File: Plugins/AudioPlayer/src/ios/AudioPlayer/AudioTOCTestament.swift
//
//  AudioTOCTestament.swift
//  AudioPlayer
//
//  Created by Gary Griswold on 8/8/17.
//  Copyright © 2017 ShortSands. All rights reserved.
//

import Utility

class AudioTOCTestament {
    
    let bible: AudioTOCBible
    let damId: String
    var booksById: Dictionary<String, AudioTOCBook>
    private var booksBySeq: Dictionary<Int, AudioTOCBook>

    init(bible: AudioTOCBible, database: Sqlite3, damId: String) {
        self.bible = bible
        self.damId = damId
        self.booksById = Dictionary<String, AudioTOCBook>()
        self.booksBySeq = Dictionary<Int, AudioTOCBook>()

        let query = "SELECT bookId, bookOrder, bookName, numberOfChapters" +
            " FROM AudioBook" +
            " WHERE damId = ?" +
            " ORDER BY bookOrder"
        do {
            let resultSet = try database.queryV1(sql: query, values: [self.damId])
            for index in 0..<resultSet.count {
                let row = resultSet[index]
                let book = AudioTOCBook(testament: self, index: index, dbRow: row)
                self.booksById[book.bookId] = book
                self.booksBySeq[book.sequence] = book
            }
        } catch let err {
            print("ERROR \(Sqlite3.errorDescription(error: err))")
        }
    }
    
    deinit {
       print("***** Deinit TOCAudioTOCBible *****") 
    }
    
    func nextChapter(reference: AudioReference) -> AudioReference? {
        let ref = reference
        let book = ref.tocAudioBook
        if (ref.chapterNum < book.numberOfChapters) {
            let next = ref.chapterNum + 1
            return AudioReference(book: ref.tocAudioBook, chapterNum: next, fileType: ref.fileType)
        } else {
            if let nextBook = self.booksBySeq[reference.sequenceNum + 1] {
                return AudioReference(book: nextBook, chapter: "001", fileType: ref.fileType)
            }
        }
        return nil
    }
    
    func priorChapter(reference: AudioReference) -> AudioReference? {
        let ref = reference
        let prior = ref.chapterNum - 1
        if (prior > 0) {
            return AudioReference(book: ref.tocAudioBook, chapterNum: prior, fileType: ref.fileType)
        } else {
            if let priorBook = self.booksBySeq[reference.sequenceNum - 1] {
                return AudioReference(book: priorBook,
                                 chapterNum: priorBook.numberOfChapters, fileType: ref.fileType)
            }
        }
        return nil
    }
    
    func getBookList() -> String {
        var array = [String]()
        for (_, book) in self.booksBySeq {
            array.append(book.bookId)
        }
        return array.joined(separator: ",")
    }
    
    func toString() -> String {
        let str = "damId=" + self.damId
        return str
    }
}
